backend/dal/Services/Concrete/ProjectReportService.cs

```diff
@@ -241,10 +241,51 @@ public void Remove(ProjectReport report)
                 .SingleOrDefault(p => p.Id == report.Id) ?? throw new KeyNotFoundException();
             if (originalReport.IsFinal) throw new InvalidOperationException($"Unable to delete FINAL project reports.");
 
+            // Remove all snapshots associated with this report.
+            var snapshots = this.Context.ProjectSnapshots.Where(s => s.SnapshotOn == originalReport.To.Value).ToArray();
+
+            this.Context.ProjectSnapshots.RemoveRange(snapshots);
             this.Context.ProjectReports.Remove(originalReport);
             this.Context.CommitTransaction();
         }
 
+        /// <summary>
+        /// Add a project and snapshots to the datasource.
+        /// This provides a way to manually generate snapshots with specific projects and historical values.
+        /// Requires a valid 'report.To' date.
+        /// </summary>
+        /// <param name="report"></param>
+        /// <param name="snapshots"></param>
+        public ProjectReport Add(ProjectReport report, IEnumerable<ProjectSnapshot> snapshots)
+        {
+            this.User.ThrowIfNotAuthorized(Permissions.ReportsSpl);
+            if (!report.To.HasValue) throw new InvalidOperationException("Argument 'report.To' must have a valid date.");
+            if (report.From.HasValue && report.From > report.To) throw new InvalidOperationException("Argument 'report.From' must be equal to or greater than 'report.To'.");
```

Comment:
I think its equal to or less than

Reply:
If the date is inverse (from > to) then throw error.